# Work order:

Analyze Berkeley Earth data on global temperature change along with Climate Watch Data Portal data on CO2 emissions evolution. There are various files from Berkeley Earth, so simply start with the one you think is most promising and first do an exploratory data analysis. Always comment on your approach, your ideas, etc. also the failures.

The central result of the data analysis should be two interesting “findings”. To do this, you create a chart that was created according to the principles of explanatory data analysis. Pay attention to the Gestalt principles, data ink ration, etc.

# Datasets:
Dataset 1.: https://www.kaggle.com/datasets/berkeleyearth/climate-change-earth-surface-temperature-data

Dataset 2.: https://www.kaggle.com/datasets/ankanhore545/carbon-dioxide-emissions-of-the-world

# Which countries in the world emit emissions and which countries are more involved and which countries are less involved?

This project looks at global emissions from 1990 to 2018, using data from https://www.kaggle.com/datasets/berkeleyearth/climate-change-earth-surface-temperature-data and https://www.kaggle.com/datasets/ankanhore545/carbon-dioxide-emissions-of-the-world. The results are interactive, using the object-oriented functions to display graphs.

In [2]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
import pycountry
import geopandas

In [3]:
df = pd.read_csv("historical_emissions.csv")

In [4]:
df

Country Data source                Sector  Gas    Unit  \
0                  World        CAIT  Total including LUCF  CO2  MtCO₂e   
1                  China        CAIT  Total including LUCF  CO2  MtCO₂e   
2          United States        CAIT  Total including LUCF  CO2  MtCO₂e   
3    European Union (27)        CAIT  Total including LUCF  CO2  MtCO₂e   
4                  India        CAIT  Total including LUCF  CO2  MtCO₂e   
..                   ...         ...                   ...  ...     ...   
190                Nauru        CAIT  Total including LUCF  CO2  MtCO₂e   
191               Tuvalu        CAIT  Total including LUCF  CO2  MtCO₂e   
192                 Niue        CAIT  Total including LUCF  CO2  MtCO₂e   
193                 Fiji        CAIT  Total including LUCF  CO2  MtCO₂e   
194                Ghana        CAIT  Total including LUCF  CO2  MtCO₂e   

         2018      2017      2016      2015      2014  ...      1999  \
0    36441.55  35588.70  35160.60  34521.91  34558.59  ...  24957.30   
1     9663.36   9367.67   9164.21   9120.27   9184.77  ...   2799.84   
2     4749.57   4581.90   4656.84   4563.52   4683.35  ...   5191.66   
3     2636.99   2692.12   2669.54   2321.61   2263.78  ...   3079.46   
4     2400.25   2267.16   2149.01   2085.38   2072.03  ...    683.00   
..        ...       ...       ...       ...       ...  ...       ...   
190      0.07      0.07      0.06      0.05      0.05  ...      0.09   
191      0.01      0.01      0.01      0.01      0.01  ...      0.01   
192      0.00      0.00      0.00      0.00      0.00  ...      0.01   
193     -0.65     -0.73     -0.81     -1.03     -1.24  ...     -1.74   
194     -8.82    -10.15    -10.82     44.53     43.77  ...     16.46   

         1998      1997      1996      1995      1994      1993      1992  \
0    24895.32  25292.92  24214.92  23890.22  23260.29  23124.00  22988.29   
1     2882.75   2779.27   2715.50   2735.48   2414.50   2294.12   2068.77   
2     5172.06   5129.29   4864.46   4708.31   4654.52   4581.76   4461.62   
3     3134.03   3142.23   3210.93   3113.68   3060.09   3072.40   3133.06   
4      618.73    600.38    555.60    519.98    466.79    431.31    409.09   
..        ...       ...       ...       ...       ...       ...       ...   
190      0.10      0.10      0.10      0.11      0.11      0.11      0.12   
191      0.01      0.01      0.01      0.01      0.01      0.01      0.01   
192      0.01      0.01      0.01      0.01      0.01      0.01      0.01   
193     -1.76     -1.74     -1.72     -1.78     -1.79     -1.79     -1.78   
194     15.98     14.18     13.98     13.60     13.34     13.07     12.96   

         1991      1990  
0    23056.40  22849.92  
1     1952.78   1823.96  
2     4389.50   4426.40  
3     3247.48   3286.44  
4      386.17    341.32  
..        ...       ...  
190      0.13      0.13  
191      0.00      0.00  
192      0.01      0.01  
193     -1.81     -1.81  
194     12.44     12.67  

[195 rows x 34 columns]

In [6]:
temp = pd.read_csv("GlobalLandTemperaturesByCity1.csv")

In [7]:
temp

dt  AverageTemperature  AverageTemperatureUncertainty  \
0        1743-11-01               6.068                          1.737   
1        1743-12-01                 NaN                            NaN   
2        1744-01-01                 NaN                            NaN   
3        1744-02-01                 NaN                            NaN   
4        1744-03-01                 NaN                            NaN   
...             ...                 ...                            ...   
8599207  2013-05-01              11.464                          0.236   
8599208  2013-06-01              15.043                          0.261   
8599209  2013-07-01              18.775                          0.193   
8599210  2013-08-01              18.025                          0.298   
8599211  2013-09-01                 NaN                            NaN   

           City      Country Latitude Longitude  
0         Århus      Denmark   57.05N    10.33E  
1         Århus      Denmark   57.05N    10.33E  
2         Århus      Denmark   57.05N    10.33E  
3         Århus      Denmark   57.05N    10.33E  
4         Århus      Denmark   57.05N    10.33E  
...         ...          ...      ...       ...  
8599207  Zwolle  Netherlands   52.24N     5.26E  
8599208  Zwolle  Netherlands   52.24N     5.26E  
8599209  Zwolle  Netherlands   52.24N     5.26E  
8599210  Zwolle  Netherlands   52.24N     5.26E  
8599211  Zwolle  Netherlands   52.24N     5.26E  

[8599212 rows x 7 columns]

In [8]:
print(temp.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8599212 entries, 0 to 8599211
Data columns (total 7 columns):
 #   Column                         Dtype  
---  ------                         -----  
 0   dt                             object 
 1   AverageTemperature             float64
 2   AverageTemperatureUncertainty  float64
 3   City                           object 
 4   Country                        object 
 5   Latitude                       object 
 6   Longitude                      object 
dtypes: float64(2), object(5)
memory usage: 459.2+ MB
None


In [9]:
nan_gtc = temp[temp['AverageTemperature'].isnull()]

In [10]:
nan_gtc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 364130 entries, 1 to 8599211
Data columns (total 7 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   dt                             364130 non-null  object 
 1   AverageTemperature             0 non-null       float64
 2   AverageTemperatureUncertainty  0 non-null       float64
 3   City                           364130 non-null  object 
 4   Country                        364130 non-null  object 
 5   Latitude                       364130 non-null  object 
 6   Longitude                      364130 non-null  object 
dtypes: float64(2), object(5)
memory usage: 22.2+ MB


Of the almost 8.6 million entries from 'temp', around 364,000 cannot be displayed. Here there are a few months that are meaningless in this case. This loss of data may be from the beginning of data recording or may have occurred not too long ago.

The possible next steps to choose from are:

1. No data is deleted from the data set. You can continue to work with the original data set, even in the future, which would not be so easy because data is missing.

2. In addition to the original data set, make a copy of the data set by removing the NaN values. The master data of the data set is retained and you can work and edit the copy of the data set as you wish and you no longer have any NaN values ​​in it.

3. In addition to the original data set, make a copy of the data set and work in this copy from a certain date so that all data before the taken date is deleted.

First you work with the 3rd selection option. If it becomes necessary, since NaN values ​​are involved, this may have to be adjusted.

In [11]:
temp_1990 = temp[temp["dt"] >= "1990-01-01"]

In [12]:
# Examination of filtered NaN values after 1990
temp_1990.isnull().sum()

dt                                  0
AverageTemperature               3070
AverageTemperatureUncertainty    3070
City                                0
Country                             0
Latitude                            0
Longitude                           0
dtype: int64

In [13]:
temp_1990.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000350 entries, 2954 to 8599211
Data columns (total 7 columns):
 #   Column                         Non-Null Count    Dtype  
---  ------                         --------------    -----  
 0   dt                             1000350 non-null  object 
 1   AverageTemperature             997280 non-null   float64
 2   AverageTemperatureUncertainty  997280 non-null   float64
 3   City                           1000350 non-null  object 
 4   Country                        1000350 non-null  object 
 5   Latitude                       1000350 non-null  object 
 6   Longitude                      1000350 non-null  object 
dtypes: float64(2), object(5)
memory usage: 61.1+ MB


Out of 1,000,350 records, 3070 still have a NaN value. Since this is only a small part of the data set (around 0.31%), the NaN values are deleted. If the value of the next lines were there, this would be a manipulation of the data set and the model, so it is easy to delete the data here.

In [14]:
temp_1990 = temp_1990.dropna(how='any', axis=0)
temp_1990.isnull().sum()

dt                               0
AverageTemperature               0
AverageTemperatureUncertainty    0
City                             0
Country                          0
Latitude                         0
Longitude                        0
dtype: int64

The column 'dt' is renamed to 'Date' and the data type is converted to 'Datetime' object. In addition, the year is separated and added to the data set as an attribute.

In [15]:
temp_1990.rename(columns={'dt' : 'Date'},inplace=True)
temp_1990['Date'] = pd.to_datetime(temp_1990['Date'])
temp_1990['Year'] = temp_1990['Date'].dt.year

In [16]:
temp_1990.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 997280 entries, 2954 to 8599210
Data columns (total 8 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   Date                           997280 non-null  datetime64[ns]
 1   AverageTemperature             997280 non-null  float64       
 2   AverageTemperatureUncertainty  997280 non-null  float64       
 3   City                           997280 non-null  object        
 4   Country                        997280 non-null  object        
 5   Latitude                       997280 non-null  object        
 6   Longitude                      997280 non-null  object        
 7   Year                           997280 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 68.5+ MB


In [17]:
temp_1990.describe().transpose()

count         mean       std       min  \
AverageTemperature             997280.0    18.420500  9.884408   -38.447   
AverageTemperatureUncertainty  997280.0     0.356137  0.197880     0.040   
Year                           997280.0  2001.343173  6.836816  1990.000   

                                      25%       50%       75%       max  
AverageTemperature               12.65175    20.805    26.296    39.156  
AverageTemperatureUncertainty     0.23200     0.309     0.423     5.181  
Year                           1995.00000  2001.000  2007.000  2013.000

# Description 'temp_1990'

The file looks at the entire world - the overall picture is therefore very much influenced by the two extremes of the polar environments and the proximity to the equator.

This circumstance results in a measuring range of -38.5°C to almost +39.2°C. The preceding explanation calls for the arithmetic mean of 18.4 °C, which has been achieved on average in every city in the world considered since 1990. A quarter of all measurements between 1990 and 2013 indicate that an equal number of cities reach an average temperature between -38.5°C and 12.7°C. Half of all data appears to be in the tolerable range between 12.7°C and 26.3°C. Meanwhile, the last 25% of the data set comes from cities where higher temperatures of up to 39.2°C were measured. Another feature is the 95% confidence interval around the average temperature. Most measurement data have an extremely high precision with this confidence interval. The maximum scattering range is ±5°C.

Due to no clear and unique primary key Date, and the fact that the file temp_1990 contains worldwide data, it is initially difficult to find a suitable visual representation. The data could be converted into a map or globe from plotly, although this is only a snapshot as a model. A time change could be achieved using a slider.

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Data columns (total 34 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Country      195 non-null    object 
 1   Data source  195 non-null    object 
 2   Sector       195 non-null    object 
 3   Gas          195 non-null    object 
 4   Unit         195 non-null    object 
 5   2018         195 non-null    float64
 6   2017         195 non-null    float64
 7   2016         195 non-null    float64
 8   2015         195 non-null    float64
 9   2014         195 non-null    float64
 10  2013         195 non-null    float64
 11  2012         195 non-null    float64
 12  2011         195 non-null    float64
 13  2010         195 non-null    float64
 14  2009         195 non-null    float64
 15  2008         195 non-null    float64
 16  2007         195 non-null    float64
 17  2006         195 non-null    float64
 18  2005         195 non-null    float64
 19  2004    

In [19]:
df.describe()

2018          2017          2016          2015          2014  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     378.715949    370.386564    366.166564    358.048769    358.536513   
std     2722.060347   2656.756306   2624.062599   2578.461926   2582.937989   
min       -8.820000    -10.150000    -10.820000   -212.220000   -213.710000   
25%        4.800000      4.320000      4.395000      3.375000      3.165000   
50%       21.920000     22.310000     22.140000     18.760000     18.770000   
75%       89.770000     88.660000     87.135000     77.345000     76.875000   
max    36441.550000  35588.700000  35160.600000  34521.910000  34558.590000   

               2013          2012          2011          2010          2009  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     356.006513    350.330000    346.323846    349.297897    333.351897   
std     2559.773990   2510.957527   2481.296970   2485.033349   2363.491694   
min     -213.110000   -203.390000   -201.170000     -6.140000     -6.450000   
25%        3.035000      3.135000      2.880000      3.350000      3.620000   
50%       18.700000     19.680000     19.450000     20.010000     19.240000   
75%       74.245000     76.515000     74.805000     72.515000     69.005000   
max    34217.180000  33620.340000  33226.010000  33328.380000  31769.890000   

       ...          1999          1998          1997          1996  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...    266.157590    266.078359    270.424872    259.915077   
std    ...   1846.768138   1843.118898   1870.838154   1791.653745   
min    ...    -27.250000    -27.510000    -32.400000    -38.300000   
25%    ...      2.700000      2.320000      2.230000      2.275000   
50%    ...     15.450000     15.510000     14.570000     14.170000   
75%    ...     64.065000     62.860000     64.715000     67.710000   
max    ...  24957.300000  24895.320000  25292.920000  24214.920000   

               1995          1994          1993          1992          1991  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     256.231436    249.665949    248.475744    247.420769    248.644462   
std     1766.707757   1719.915887   1709.447361   1698.507027   1703.334650   
min      -41.950000    -45.540000    -48.040000    -52.980000    -53.980000   
25%        2.110000      2.220000      2.175000      2.260000      2.775000   
50%       14.130000     13.830000     14.640000     14.120000     14.610000   
75%       62.015000     61.870000     59.820000     59.120000     63.080000   
max    23890.220000  23260.290000  23124.000000  22988.290000  23056.400000   

               1990  
count    194.000000  
mean     248.028402  
std     1693.633278  
min      -53.880000  
25%        2.590000  
50%       15.295000  
75%       61.997500  
max    22849.920000  

[8 rows x 29 columns]

# Descriptions of the existing data (name, number, data type)

In [20]:
df.value_counts("Data source")

Data source
CAIT    195
dtype: int64

In [21]:
df.value_counts("Gas")

Gas
CO2    195
dtype: int64

In [22]:
df.value_counts("Sector")

Sector
Total including LUCF    195
dtype: int64

In [23]:
df.value_counts("Unit")

Unit
MtCO₂e    195
dtype: int64

In [24]:
df.drop(["Data source", "Sector", "Unit", "Gas"], axis=1, inplace=True)

In [25]:
# Eliminate lines for the world and the European Union
df = df[df["Country"] != "World"]
df = df[df["Country"] != "European Union (27)"]

In [30]:
# Creates a bar chart showing the countries with the most emissions for each year

def most_emission(df):
    
    # Set countries as index
    df = df.set_index("Country")
    
    # contains the charts to be displayed
    item_list = []
    
    # Reverse the list so that 1990 appears first
    for col in df.columns.tolist()[::-1]:
        # sort df
        df = df.sort_values(col, ascending=False).head(10)
        # create the chart
        item = dict(args=[{"x" : [df.index], "y" : [df[col]] }, # Set x and y axes
                         {"title" : "Countries with the most emissions in " + col} # Set title
                         ],
                         label=col, method="update") # label = Button name
                                                     # method="update" changes data and layout
        item_list.append(item)
    
    # Creates the first chart to match the first value of our menu
    firstviz = df.sort_values(df.columns.tolist()[::-1][0], ascending=False).head(10)
    
    fig = px.bar(firstviz, x=firstviz.index, y=firstviz.columns.tolist()[::-1][0],
                 color_discrete_sequence=["#4ADEDE"], template="plotly_dark",
                 text_auto=True,
                 title="Countries with the most emissions in " + df.columns.tolist()[::-1][0])
    # title
    fig.update_layout(title_x=0.5)

    # Create menu
    fig.update_layout(
            updatemenus=[
                # item list = List of charts
                dict(buttons=list(item_list),  

                 # Location of the drop-down menu
                direction="down",showactive=True,x=0.005,
                xanchor="left",y=1.4,yanchor="top") ] )
    
    return fig.show()

In [31]:
most_emission(df)

In [32]:
# Function to plot CO2 emissions by year for a country, with a line for global mean emissions

def emission_country(country):
    # df for the selected country
    df_country = df[df["Country"] == country].drop("Country", axis=1).T.iloc[::-1]
    
    # Column name = index, replace it with the country name
    column_name = df_country.columns[0]
    column_dict = {}
    column_dict[column_name] = country
    df_country.rename(columns=column_dict, inplace=True)
    
    # Maintains the world's average emission
    world_mean = df.describe().loc["mean"].iloc[::-1]    

    # Create a world average line graph
    fig = px.line(world_mean, x=world_mean.index, y=world_mean.values,
                 color_discrete_sequence=["red"],
                 title=f"Entwicklung der CO2-Emission für {country}",
                 template="plotly_dark")
    
    # Creates a bar chart for the selected country
    fig2 = px.bar(df_country, x=df_country.index, y=country,
                 color_discrete_sequence=["#4ADEDE"],
                 text_auto=True)
    
    # Adds the bar chart to the first figure
    fig.add_trace(fig2.data[0])
    
    # title
    fig.update_layout(title_x=0.5)
    # Rotate checkmark labels
    fig.update_xaxes(tickangle=45)
    # Hide Y axis
    fig.update_yaxes(visible=False)
    
    return fig.show()

# Assumption for the countries France, Germany and China:

In France and Germany, emissions will be relatively low and the world average will fall from 1990 to 2018. Since the emissions of the two countries are relatively low compared to the rest of the world.
When it comes to China, I can well imagine that emissions will be high and that the world average will rise from 1990 to 2018.

In [34]:
# CO2 emissions per year for France. The red line is the global average emission.
# You can see that it has been slowly increasing over the years; in France you can see a slight decline in emissions.
emission_country("France")

In [35]:

# CO2 emissions per year for Germany. The red line is the global average emission.
# You can see that it has been slowly increasing over the years; in Germany you can see a very slight decline in emissions.
emission_country("Germany")

In [36]:
emission_country("China")
# The emissions emitted in China are so high relative to the rest of the world that the red line in this graph is straight.

# Conclusion on the assumption for the countries France, Germany and China

My assumption has only been partially confirmed. It was true that the two countries France and Germany are relatively low, but that the world average would fall was wrong. The world average has increased. I can explain this by the fact that in the period up to 2018 more and more emissions will be emitted by more industries etc., so that the global average will rise instead of falling.

This time I made the wrong assumption about China. In China, the emissions load was still relatively low in 1990, but over the years the emissions load has continued to rise, so that the world average is not increasing, but is a straight line due to the global emissions load.

In [37]:
df_evolution = df.copy()

# Calculate development between 1990 and 2018
df_evolution["Evolution"] = df_evolution["2018"] - df_evolution["1990"]

In [38]:
df_worst_evolution = df_evolution.sort_values("Evolution", ascending=False).head(10)

fig = px.bar(df_worst_evolution,x="Country", y="Evolution", template="plotly_dark",
             color_discrete_sequence=["#4ADEDE"], text_auto=True,
             title="Countries with the largest increase in CO2 emissions between 1990 and 2018")
fig.update_layout(title_x=0.5)
fig.show()

# Increase in emissions from 1990 to 2018

China is in first place with the strongest increase in emissions. India comes in second place, there is a very big difference to China, with more than 5.5 million emissions. From 10th to 3rd place, everyone is increasing steadily for a long time, but has many times fewer emissions than China.

In [39]:
df_best_evolution = df_evolution.sort_values("Evolution", ascending=True).head(10)

fig = px.bar(df_best_evolution,x="Country", y="Evolution", template="plotly_dark",
             color_discrete_sequence=["#4ADEDE"], text_auto=True,
             title="Countries with the largest declines in CO2 emissions between 1990 and 2018")
fig.update_layout(title_x=0.5)
fig.show()

# Decrease in emissions from 1990 to 2018

Russia is in first place with the largest decline in emissions. Coming in 2nd and 3rd place are Ukraine and Brazil, who don't have much difference from each other, but Russia and Germany, who are in 4th place, is a big jump. Places 5 to 10 are constantly decreasing for all countries.

In [40]:
# Add a "CODE" column with the 3 letter country code
# I got the function here: https://melaniesoek0120.medium.com/data-visualization-how-to-plot-a-map-with-geopandas-in-python-73b10dcd4b4b
def alpha3code(column):
    CODE=[]
    for country in column:
        try:
            code=pycountry.countries.get(name=country).alpha_3
           # .alpha_3 means 3 letter country code
            CODE.append(code)
        except:
            CODE.append('None')
    return CODE
# Create a column for code
df['CODE']=alpha3code(df.Country)
df.head()

Country     2018     2017     2016     2015     2014     2013  \
1          China  9663.36  9367.67  9164.21  9120.27  9184.77  9226.51   
2  United States  4749.57  4581.90  4656.84  4563.52  4683.35  4670.34   
4          India  2400.25  2267.16  2149.01  2085.38  2072.03  1901.98   
5      Indonesia  1269.55  1081.11  1068.10  1574.76  1539.23  1250.27   
6          Japan  1074.08  1119.86  1139.33  1131.74  1170.11  1211.27   

      2012     2011     2010  ...     1998     1997     1996     1995  \
1  8823.05  8568.09  8138.34  ...  2882.75  2779.27  2715.50  2735.48   
2  4531.16  4753.12  4990.96  ...  5172.06  5129.29  4864.46  4708.31   
4  1843.74  1695.97  1670.29  ...   618.73   600.38   555.60   519.98   
5  1303.87  1285.69   773.92  ...   907.96  1484.56   772.38   909.21   
6  1202.66  1162.67  1057.96  ...  1055.50  1098.26  1110.51  1097.10   

      1994     1993     1992     1991     1990  CODE  
1  2414.50  2294.12  2068.77  1952.78  1823.96   CHN  
2  4654.52  4581.76  4461.62  4389.50  4426.40   USA  
4   466.79   431.31   409.09   386.17   341.32   IND  
5   881.31   869.88   855.30   846.23   832.39   IDN  
6  1087.22  1033.99  1041.79  1029.99  1016.89   JPN  

[5 rows x 31 columns]

In [41]:
df_world = df.copy()
df_world.value_counts("CODE")

CODE
None    18
AFG      1
NRU      1
NAM      1
NER      1
        ..
GNB      1
GNQ      1
GRC      1
GRD      1
ZWE      1
Length: 176, dtype: int64

In [42]:
# There are countries with missing code

# Country list without code

missing_country = df_world[df_world["CODE"] == "None"].Country.tolist()

# Get the corresponding code from the Internet
missing_code = ["RUS", "IRN", "COD", "KOR", "VNM", "VEN", "CZE", "BOL", "TZA", "SYR", "LAO", "PRK", "COG", "MDA", "BRN", "MKD", "CPV", "FSM"]

# zip
country_code = list(zip(missing_country, missing_code))

# Add the correct code to the corresponding country
for i,j in country_code:
    df_world.loc[df_world.Country == i, "CODE"] = j

In [43]:
# Function to display the map in which we can select the year we want

def world_map(df, year):
    
    # Creation of the map
    # Locations = Alpha 3 code for each country
    
    fig = px.choropleth(df_world, locations=df_world.CODE, color=year,
                        color_continuous_scale='RdYlGn_r', template="plotly_dark",
                       title="Global CO2 emissions in " + year)
    
    # set the mapbox
    # Adjust the zoom when the chart is displayed
    fig.update_layout(mapbox_style="carto-positron",
                      mapbox_zoom=1)
    # Title
    fig.update_layout(title_x=0.5, margin={"r":0, "t":30,"l":0,"b":0})
    return fig.show()

# Assumption about the world maps:
-> I could imagine that the ones that were already red in 1990 would stay red because the industry will certainly continue to be in demand and will continue to operate.

-> I could also imagine that over time one or the other country will move towards red, but there won't be a big change.

In [44]:
#Status quo of emissions around 1990.
#Here you can see that the emissions burden in the USA is very high.
#Furthermore, in China you can see an increased emissions value (yellow), but not yet so high that it is shown in red here.
#In the rest of the world, the emissions burden is green, but that does not mean that there is no emissions burden in these countries, just that it is relatively low.

world_map(df_world, "1990")

In [45]:
# There is more green, but the scale has changed. The countries in yellow, in 1990, may have the same amount of CO2 emissions and are in green for 2018.

world_map(df_world, "2018")

# Conclusion on my assumption
I was wrong in my assumption. The country USA, which was red in 1990, is yellow in 2018, which means relegation. Brazil, which went from yellow to green, has also reduced its emissions. The exact opposite of the USA happened in China, they were yellow but in 2018 they are red. In the case of Russia, you could previously see in the graphics that there was a sharp decline, from 1990 yellow to 2018 green.

-> The USA reduced emissions between 1990 (red) and 2018 (yellow).

-> China made massive gains in the period from 1990 (yellow) to 2018 (red).

-> Meanwhile, all other countries are equally loved or have been able to reduce their emissions.

# CONCLUSION
In industrialized nations, reducing CO2 emissions has been less successful.

The last two cards show two different worlds. In 1990 the US was to blame for a large amount of CO2 emissions and 28 years later, it is China. China's emissions increased almost exponentially between 1990 and 2018. Added together, China is in first place and produced almost as much CO2 as places 2 to 10 combined and there is still room for improvement with China.

The temperature cannot be linked to the emissions, but the other way around. Due to the release of emissions, the temperature in the world is rising, but this is another issue (climate change).

In summary: Which countries contribute to emissions.

China had the largest emissions in the world in 2018, but was initially still in the green. Russia had the largest decline in emissions in 20198, was initially in the red but has improved the most. Most of the other countries that were green or yellow in 1990 are still green or have become green in 2018 because, in comparison, they have significantly lower emissions.